File: sdk/typescript/packages/core/src/judge.ts
/**
 * TOLAP Semantic Judge (canonical spec §15.4)
 *
 * A semantic check on whether a tool call serves the purpose it was authorized for,
 * plus the gate that turns a policy's `purposeProfile.judge` block into an actual
 * invocation.
 *
 * **Advisory and strictly subtractive.** A judge runs only after the three
 * deterministic checks — purpose filtering at resolution, action validation, chain
 * validation — have allowed a call, and it can only take that allowance away. It can
 * never permit something they denied, because a non-deterministic component must not
 * be able to widen access: otherwise a prompt that talks the model round becomes a
 * privilege escalation.
 *
 * The gate exists because a policy can configure a judge in full — a model, a
 * history window, two thresholds, a latency budget — and none of it takes effect
 * unless something reads those values and applies them. Left to each integrator, the
 * predictable outcome is a deployment where the judge runs with a window and
 * thresholds nobody chose and where the policy's `model` is quietly ignored: a
 * control the configuration implies and that never runs.
 *
 * The wrappers call `evaluateJudge` for you when a judge is configured: `preExecuteAsync`
 * runs the deterministic checks and then the gate, so a policy's judge block applies
 * without glue of yours. Call it directly only if you are not using a wrapper — and render
 * the call with `renderToolCall` if you do, so your history and a wrapper's stay
 * comparable. An implementation that calls a model lives in `@aws/tolap-mcp` (see
 * `BedrockJudge`); `@aws/tolap-core` ships with no runtime dependencies and a judge needs a
 * transport.
 */

import type {
  EffectivePolicy,
  JudgeConfig,
  PurposeProfile,
} from "./types.js";
import type { ToolCallHistory } from "./history.js";

// ---------------------------------------------------------------------------
// Documented defaults (applied at READ time, never stored on a JudgeConfig)
// ---------------------------------------------------------------------------
//
// `JudgeConfig`'s fields are all optional and these are applied when a field is
// read. Storing them on the object instead would serialize values the author never
// wrote, changing the canonical bytes -- and so the signature -- of every
// purpose-bound policy. The numbers match `schema/v1.0/*.schema.json`'s `default`
// keywords, and `schema-conformance`-style tests pin that agreement.

/** Confidence at or above which a verdict is final, when unconfigured. */
export const DEFAULT_CONFIDENCE_THRESHOLD = 0.85;

/** Confidence below which a call escalates, when unconfigured. */
export const DEFAULT_ESCALATION_THRESHOLD = 0.6;

/** Wall-clock budget for one evaluation, when unconfigured. */
export const DEFAULT_MAX_LATENCY_MS = 2000;

/** Number of preceding calls to supply, when unconfigured. */
export const DEFAULT_HISTORY_WINDOW = 10;

/**
 * The escalation reason when the judge is not the model the policy asked for.
 *
 * Part of the contract; integrators log and branch on it. Escalation rather than a
 * hard block: the call may be perfectly legitimate and the fault is in the
 * deployment, so a human is the right destination. What it must never be is a silent
 * substitution.
 *
 * `JudgeOutcome.reason` carries this as a **prefix**, followed by the two model ids —
 * a mismatch is a configuration fault and the only useful log line names both sides.
 * Branch with `reason.startsWith(JUDGE_MODEL_MISMATCH_REASON)`, never on equality.
 */
export const JUDGE_MODEL_MISMATCH_REASON = "judge model mismatch";

/** The reason a policy that asked for no judge reports. */
export const NO_JUDGE_CONFIGURED_REASON = "no judge configured";

/**
 * The escalation reason when the judge itself threw.
 *
 * §15.4 requires a timeout, a transport failure and an unparseable response to "produce
 * escalation rather than an exception". `BedrockJudge` honours that internally, but {@link Judge}
 * is a public interface and a custom implementation can throw — which would put an exception on
 * the authorization path, and the natural fix for that at a call site is a `catch` that returns
 * "allow". Caught in {@link evaluateJudge} so the safe reading is the default rather than
 * something each integrator has to remember. Byte-identical to .NET's
 * `JudgeGate.JudgeFailedReason` and Python's `JUDGE_FAILED_REASON`.
 */
export const JUDGE_FAILED_REASON = "judge invocation failed";

// ---------------------------------------------------------------------------
// The judge contract
// ---------------------------------------------------------------------------

/** What a judge is asked to decide. */
export interface JudgeRequest {
  /**
   * The profile from the resolved policy. Its `description` is what gives the judge
   * something to compare the call against, which is why an author who enables the
   * judge should write one.
   */
  purpose: PurposeProfile;
  /** A rendering of the call under consideration. */
  currentToolCall: string;
  /**
   * Preceding calls, **oldest first**, bounded by `JudgeConfig.historyWindow`.
   * Drift is a property of a sequence rather than of one call: an agent asking for
   * one more column each turn looks reasonable at every single step.
   */
  recentHistory: string[];
  /** The wall-clock budget. Exceeding it escalates rather than allowing. */
  maxLatencyMs: number;
}

/** A judge's verdict. */
export interface JudgeResult {
  /** Whether the call serves the declared purpose. */
  aligned: boolean;
  /**
   * How sure the judge is, in `[0, 1]`. A value outside that range is unusable and
   * escalates — see {@link getDisposition}.
   */
  confidence: number;
  /** The judge's explanation, for the audit trail. */
  reasoning: string;
  /** Optional machine-readable markers for downstream triage. */
  flags?: string[];
}

/** What to do with a judge's verdict. */
export enum JudgeDisposition {
  /** Proceed. The deterministic checks already passed. */
  Allow = "allow",
  /** Refuse. The judge is confident the call does not serve the purpose. */
  Block = "block",
  /**
   * Refer to a human. **Not** an allow: a wrapper with no escalation handler must
   * deny, or "escalate to review" silently means "permit" wherever the review step
   * was never built.
   */
  Escalate = "escalate",
}

/**
 * What {@link evaluateJudge} decided, and why.
 *
 * A bare {@link JudgeDisposition} is not enough. `escalate` is returned both when the
 * deployment wired up the wrong model and when the judge genuinely could not tell, and
 * those call for entirely different responses — fix the configuration versus route to a
 * reviewer. Returning only the disposition left {@link JUDGE_MODEL_MISMATCH_REASON}
 * unreachable from the call site, which made a mandated reason string decoration.
 */
export interface JudgeOutcome {
  disposition: JudgeDisposition;
  /**
   * Why. The judge's own `reasoning` on a real verdict; otherwise a reason this module
   * owns — {@link NO_JUDGE_CONFIGURED_REASON}, or a string beginning with
   * {@link JUDGE_MODEL_MISMATCH_REASON}.
   */
  reason: string;
  /**
   * The verdict, or **absent when no model was consulted** — either because the policy
   * asked for no judge, or because the model did not match and the call was refused
   * before being issued. Absent is therefore a positive statement: no tokens were
   * spent and nothing a model said is being reported.
   */
  result?: JudgeResult;
  /**
   * Whether the call may proceed. **`escalate` is not allowed.**
   *
   * Precomputed rather than left to the caller, because a caller writing
   * `disposition !== Block` is the most likely route to "escalate to human review"
   * quietly meaning "permit" in a deployment that never built the review step.
   */
  allowed: boolean;
}

/** Build a {@link JudgeOutcome}, deriving `allowed` so no caller has to. */
function outcome(
  disposition: JudgeDisposition,
  reason: string,
  result?: JudgeResult,
): JudgeOutcome {
  return {
    disposition,
    reason,
    ...(result === undefined ? {} : { result }),
    // The one place `escalate` is turned into a boolean, so there is one place to read
    // when asking whether it counts as permission. It does not.
    allowed: disposition === JudgeDisposition.Allow,
  };
}

/**
 * A semantic judge.
 *
 * Named `Judge` rather than `IJudge` to match this SDK's convention — `PolicyStore`
 * and `ReplayGuard` are the TypeScript spellings of .NET's `IPolicyStore` and
 * `IReplayGuard`. The .NET counterpart is `Tolap.Core.IJudge`.
 */
export interface Judge {
  /**
   * The model this judge actually invokes.
   *
   * Declared so {@link evaluateJudge} can check it against the
   * `JudgeConfig.model` the policy asked for. Without it, a policy demanding one
   * model would be silently judged by whatever the deployment happened to wire up
   * — the policy field would look like a control and be decoration. A verdict is
   * only meaningful against the model that produced it, which is also why two
   * policies naming different models refuse to merge.
   *
   * Reported by the implementation rather than configured alongside it, so the
   * value comes from whatever really issues the call.
   */
  readonly modelId: string;

  /**
   * Evaluate a call against its purpose.
   *
   * Implementations should surface a timeout or a transport failure as a
   * low-confidence result (or, for caller cancellation, by rejecting) rather than
   * a confident `aligned: true`. The caller maps either to
   * {@link JudgeDisposition.Escalate}; a fabricated confident alignment is the one
   * response that cannot be recovered from.
   *
   * @param signal Caller cancellation, distinct from the policy's latency budget.
   */
  evaluate(request: JudgeRequest, signal?: AbortSignal): Promise<JudgeResult>;
}

// ---------------------------------------------------------------------------
// Verdict -> disposition
// ---------------------------------------------------------------------------

/**
 * Decide what to do with a verdict.
 *
 * Every path that is not a confident verdict escalates, and escalation denies
 * unless a handler is wired. The ordering matters: the unusable-input checks come
 * **first**, so a malformed result cannot reach the threshold comparisons and win
 * one.
 *
 * Three unusable-input cases, each of which would otherwise produce a confident
 * allow:
 *
 * - A confidence outside `[0, 1]`. A judge reporting `1.5` has malfunctioned, and
 *   comparing it against a threshold would grant it more authority than a correct
 *   answer.
 * - Thresholds **inverted**, with escalation above confidence. There is no reading
 *   of that configuration to act on, so it escalates rather than picking whichever
 *   bound happens to be checked first.
 * - `NaN`, which fails every comparison and would otherwise fall through to
 *   whatever the last branch happened to be. In JavaScript this matters more than
 *   elsewhere: `NaN < 0.6` and `NaN > 1` are both false, so a bare comparison chain
 *   ends at the confident arm.
 */
export function getDisposition(
  result: JudgeResult,
  config?: JudgeConfig,
): JudgeDisposition {
  const confidenceThreshold =
    config?.confidenceThreshold ?? DEFAULT_CONFIDENCE_THRESHOLD;
  const escalationThreshold =
    config?.escalationThreshold ?? DEFAULT_ESCALATION_THRESHOLD;

  if (
    Number.isNaN(result.confidence) ||
    result.confidence < 0 ||
    result.confidence > 1
  ) {
    return JudgeDisposition.Escalate;
  }

  if (escalationThreshold > confidenceThreshold) {
    return JudgeDisposition.Escalate;
  }

  if (result.confidence < escalationThreshold) {
    return JudgeDisposition.Escalate;
  }

  if (result.confidence < confidenceThreshold) {
    return JudgeDisposition.Escalate;
  }

  return result.aligned ? JudgeDisposition.Allow : JudgeDisposition.Block;
}

// ---------------------------------------------------------------------------
// The gate
// ---------------------------------------------------------------------------

/**
 * Whether this policy asks for a judge at all.
 *
 * Checked against `=== true` rather than for truthiness, because
 * `JudgeConfig.enabled` is optional: absent means "not configured", which is not
 * the same statement as an explicit `false` even though both mean no judge today.
 */
export function judgeEnabled(policy: EffectivePolicy): boolean {
  return policy.purposeProfile?.judge?.enabled === true;
}

/**
 * How many preceding calls the policy wants the judge to see.
 *
 * Read this to size a `ToolCallHistory`, so the window is the policy author's
 * choice rather than the integrator's default. A window smaller than the policy
 * asked for hides exactly the trajectory the judge was enabled to notice.
 */
export function judgeHistoryWindow(policy: EffectivePolicy): number {
  return policy.purposeProfile?.judge?.historyWindow ?? DEFAULT_HISTORY_WINDOW;
}

/**
 * Build the judge request for a call, from the policy's judge configuration.
 *
 * @returns
 * The request, or `undefined` when this policy has no judge enabled — in which case
 * there is nothing to ask and the deterministic decision stands.
 *
 * The latency budget comes from the policy rather than from the caller, which is
 * the whole point: `maxLatencyMs` is a policy field and a caller passing its own
 * value would make it advisory.
 *
 * @throws Error when `currentToolCall` is not a string. Guarded at runtime because
 * the alternative is a prompt in which the call under review reads as the literal
 * `undefined`, and a judge cannot report that as suspicious.
 */
export function buildJudgeRequest(
  policy: EffectivePolicy,
  currentToolCall: string,
  history?: ToolCallHistory,
): JudgeRequest | undefined {
  if (typeof currentToolCall !== "string") {
    throw new Error(
      `buildJudgeRequest requires a string rendering of the call, received ` +
        `${currentToolCall === null ? "null" : typeof currentToolCall}`,
    );
  }

  if (!judgeEnabled(policy)) return undefined;

  // Non-null after `judgeEnabled`: it is exactly the assertion that both are set.
  const profile = policy.purposeProfile as PurposeProfile;
  const judge = profile.judge as JudgeConfig;

  // Trimmed to the policy's window even when the caller's history is larger, so an
  // oversized buffer cannot quietly widen what the policy chose to send.
  const window = judge.historyWindow ?? DEFAULT_HISTORY_WINDOW;
  const recorded = history?.getRecent() ?? [];
  const recentHistory =
    recorded.length > window ? recorded.slice(recorded.length - window) : recorded;

  return {
    purpose: profile,
    currentToolCall,
    recentHistory,
    maxLatencyMs: judge.maxLatencyMs ?? DEFAULT_MAX_LATENCY_MS,
  };
}

/**
 * Whether the judge in hand is the one the policy asked for.
 *
 * A policy naming no model accepts any judge: the field is optional, and requiring
 * it would make every judge-enabled policy fail until someone pinned a model id
 * that differs per account and region.
 *
 * When a policy does name one, the comparison is **exact and case-sensitive**. Not
 * a prefix or substring match: `claude-sonnet` and `claude-sonnet-5` are different
 * models, and a prefix rule would let a deployment satisfy a policy demanding one
 * by wiring the other. This mirrors the `purposeId` comparison — an identifier is
 * matched, not a pattern.
 */
function modelMatches(configured: string | undefined, actual: string): boolean {
  return configured === undefined || configured === actual;
}

/**
 * Run the judge for a call and map the verdict to a disposition.
 *
 * @param policy The resolved policy. No judge enabled means
 * {@link JudgeDisposition.Allow}.
 * @param judge The judge. Its `modelId` is verified **first**.
 * @param currentToolCall A rendering of the call under consideration.
 * @param history Preceding calls; trimmed to the policy's window.
 * @param signal Caller cancellation. Distinct from the policy's latency budget.
 *
 * @returns
 * A {@link JudgeOutcome}: what to do, why, and the verdict when one was obtained.
 * `JudgeDisposition.Escalate` is **not** an allow — read `outcome.allowed`, which
 * says so — or "escalate to review" means "permit" in every deployment that never
 * built one.
 *
 * The model check runs before the call, not after. Invoking the wrong model and
 * then noticing would have spent the tokens and, worse, produced a verdict that
 * reads as authoritative in an audit log. That escalation's reason begins with
 * {@link JUDGE_MODEL_MISMATCH_REASON} and names both models, and its `result` is
 * absent because nothing was asked.
 *
 * @throws Error when `judge` is nullish.
 */
export async function evaluateJudge(
  policy: EffectivePolicy,
  judge: Judge,
  currentToolCall: string,
  history?: ToolCallHistory,
  signal?: AbortSignal,
): Promise<JudgeOutcome> {
  if (judge === null || judge === undefined) {
    throw new Error("evaluateJudge requires a judge; received none");
  }

  const request = buildJudgeRequest(policy, currentToolCall, history);
  if (request === undefined) {
    return outcome(JudgeDisposition.Allow, NO_JUDGE_CONFIGURED_REASON);
  }

  const config = (policy.purposeProfile as PurposeProfile).judge as JudgeConfig;

  if (!modelMatches(config.model, judge.modelId)) {
    return outcome(
      JudgeDisposition.Escalate,
      // Both sides named. A log line saying only "judge model mismatch" tells an
      // operator that something is misconfigured and nothing about what to change.
      `${JUDGE_MODEL_MISMATCH_REASON}: policy requires '${config.model}', ` +
        `judge reports '${judge.modelId}'`,
    );
  }

  let result: JudgeResult;
  try {
    result = await judge.evaluate(request, signal);
  } catch (error) {
    // Deliberately broad: `Judge` is a public interface, so the thrown value is whatever an
    // implementation produces, and every one of them means the same thing here — no verdict.
    // A caller's own abort is re-thrown so it stays distinguishable from a judge failure.
    if (signal?.aborted) throw error;
    const name = error instanceof Error ? error.constructor.name : typeof error;
    return outcome(JudgeDisposition.Escalate, `${JUDGE_FAILED_REASON}: ${name}`);
  }

  // The judge's own words become the reason, so a `block` or an ambiguous `escalate`
  // carries the explanation the model gave rather than a reason this module invented.
  return outcome(getDisposition(result, config), result.reasoning, result);
}

File: sdk/python/tolap-core/tolap_core/judge.py
"""The optional semantic judge (canonical-enforcement-spec.md section 15.4).

Advisory and strictly subtractive. A judge runs only after the three deterministic
checks -- purpose filtering, action validation, chain validation -- have allowed a
call, and it can only take that allowance away. It can never permit something they
denied: a non-deterministic component must not be able to widen access, because then
a prompt that talks the model round becomes a privilege escalation.

The interface lives here in ``tolap_core`` and every implementation that calls a model
lives in ``tolap_mcp``, which already carries external dependencies. Core ships with
none, and a judge needs a network client.

The three functions below are this SDK's equivalent of .NET's ``JudgeGate``:
:func:`judge_enabled`, :func:`judge_history_window`, :func:`build_judge_request` and
:func:`evaluate_judge`. They exist because a policy can configure a judge in full -- a
model, a history window, two thresholds, a latency budget -- and none of it takes
effect unless something reads those values and applies them. Left to each integrator,
the predictable outcome is a deployment where the judge runs with a window and
thresholds nobody chose, and where the policy's ``model`` is quietly ignored: a
control the configuration implies and that never runs.

The wrappers deliberately do **not** call any of this. The judge is opt-in glue: it
needs a network client, an escalation destination and a retention decision about
tool-call history, none of which a wrapper can assume.
"""

from __future__ import annotations

import math
from abc import ABC, abstractmethod
from dataclasses import dataclass
from enum import Enum
from typing import TYPE_CHECKING

from tolap_core.models import EffectivePolicy, JudgeConfig, PurposeProfile

if TYPE_CHECKING:  # pragma: no cover - import cycle broken for typing only
    from tolap_core.history import ToolCallHistory


DEFAULT_CONFIDENCE_THRESHOLD = 0.85
"""Confidence at or above which a verdict is final, when unconfigured."""

DEFAULT_ESCALATION_THRESHOLD = 0.60
"""Confidence below which a call escalates, when unconfigured."""

DEFAULT_MAX_LATENCY_MS = 2000
"""Wall-clock budget for one evaluation, when unconfigured."""

DEFAULT_HISTORY_WINDOW = 10
"""Number of preceding calls to supply, when unconfigured."""

JUDGE_MODEL_MISMATCH_REASON = "judge model mismatch"

#: The reason on an allow that no judge was asked for.
#:
#: Byte-identical to .NET's ``JudgeGate.NoJudgeConfiguredReason`` and TypeScript's
#: ``NO_JUDGE_CONFIGURED_REASON``. It existed only in TypeScript, was an inline literal
#: spelled differently in .NET, and was absent here -- three behaviours for one outcome in a
#: field integrators log and branch on.
NO_JUDGE_CONFIGURED_REASON = "no judge configured"

#: The escalation reason when the judge itself raised.
#:
#: Section 15.4 requires a timeout, a transport failure and an unparseable response to
#: "produce escalation rather than an exception". ``BedrockJudge`` honours that internally,
#: but :class:`Judge` is a public ABC and a custom implementation can raise -- which would put
#: an exception on the authorization path, and the natural fix for that at a call site is an
#: ``except`` that returns "allow". Caught in :func:`evaluate_judge` so the safe reading is the
#: default rather than something each integrator has to remember.
#: Byte-identical to .NET's ``JudgeGate.JudgeFailedReason`` and TypeScript's
#: ``JUDGE_FAILED_REASON``.
JUDGE_FAILED_REASON = "judge invocation failed"
"""The escalation reason when the judge is not the model the policy asked for.

Part of the contract, and spelled as TypeScript's ``JUDGE_MODEL_MISMATCH_REASON`` and
.NET's ``JudgeGate.ModelMismatchReason`` spell it -- three spellings of one constant is
drift, whatever each language's own conventions say about interfaces.

Escalation rather than a hard block: the call may be perfectly legitimate, and the fault
is in the deployment, so a human is the right destination. What it must never be is a
silent substitution.

:attr:`JudgeOutcome.reason` carries this as a **prefix**, followed by the two model ids,
so an integrator can branch on the constant and still read which model was expected and
which was wired.
"""


@dataclass
class JudgeRequest:
    """What a judge is asked to decide.

    ``purpose`` is the profile from the resolved policy. Its ``description`` is what
    gives the judge something to compare the call against, which is why an author who
    enables the judge should write one.

    ``recent_history`` holds preceding calls, oldest first, bounded by
    ``JudgeConfig.history_window``. Drift is a property of a sequence rather than of
    one call: an agent asking for one more column each turn looks reasonable at every
    single step.

    ``max_latency_ms`` is the wall-clock budget. Exceeding it escalates rather than
    allowing.
    """

    purpose: PurposeProfile
    current_tool_call: str
    recent_history: list[str]
    max_latency_ms: int


@dataclass
class JudgeResult:
    """A judge's verdict.

    ``confidence`` is how sure the judge is, in ``[0, 1]``. A value outside that range
    is unusable and escalates -- see :func:`judge_disposition`.

    ``reasoning`` is the judge's explanation, for the audit trail. ``flags`` are
    optional machine-readable markers for downstream triage.
    """

    aligned: bool
    confidence: float
    reasoning: str
    flags: list[str] | None = None


class JudgeDisposition(Enum):
    """What to do with a judge's verdict."""

    #: Proceed. The deterministic checks already passed.
    allow = "allow"
    #: Refuse. The judge is confident the call does not serve the purpose.
    block = "block"
    #: Refer to a human. **Not** an allow: a wrapper with no escalation handler must
    #: deny, or "escalate to review" silently means "permit" wherever the review step
    #: was never built.
    escalate = "escalate"


@dataclass(frozen=True)
class JudgeOutcome:
    """What :func:`evaluate_judge` decided, and why.

    A bare :class:`JudgeDisposition` is not enough. ``escalate`` is returned both when a
    deployment wired the wrong model and when the judge genuinely could not tell, and those
    call for entirely different responses -- one is a configuration fault to fix, the other
    is a call to put in front of a human. Without a reason the mandated
    :data:`JUDGE_MODEL_MISMATCH_REASON` is unreachable from the call site, so the policy's
    ``model`` field would be checked and the check would be invisible.

    ``result`` is ``None`` whenever no model was consulted: either the policy enables no
    judge, or the model did not match and the call was refused before being issued. That
    distinction is what tells an audit trail "no verdict exists" apart from "a verdict was
    obtained and it was uncertain".

    ``reason`` is populated on **every** disposition, including an allow -- deliberately
    unlike :class:`~tolap_core.enforcement.AccessResult`, where an allow carries none. The
    convention does not transfer, because the two fields answer different questions. An
    ``AccessResult`` allow has nothing to explain: the policy permitted the call and that is
    the whole story. A judge allow does: a model formed a view, and an audit trail reviewing a
    borderline approval wants that view in the record rather than having to reach into
    ``result`` for it -- or, when no judge ran at all, wants to see that stated rather than
    inferring it from an empty field. All three SDKs populate it identically.
    """

    disposition: JudgeDisposition
    reason: str = NO_JUDGE_CONFIGURED_REASON
    result: JudgeResult | None = None

    @property
    def allowed(self) -> bool:
        """Whether the call may proceed.

        ``escalate`` is **not** allowed. A caller with a human-review path branches on
        :attr:`disposition` instead; a caller without one gets the fail-closed reading by
        default, rather than "escalate to review" silently meaning "permit" wherever the
        review step was never built.
        """
        return self.disposition is JudgeDisposition.allow


class Judge(ABC):
    """A semantic check on whether a tool call serves its authorized purpose.

    Named without an ``I`` prefix, matching this SDK's other seams
    (:class:`~tolap_core.context.ReplayGuard`,
    :class:`~tolap_store.PolicyStore`) where .NET writes ``IReplayGuard`` and
    ``IPolicyStore``.

    Synchronous, unlike .NET's ``IJudge.EvaluateAsync``, because every other
    enforcement entry point in this SDK is synchronous -- the HTTP wrapper drives a
    blocking ``httpx.Client``. An async judge would make the one optional check the
    only reason to introduce an event loop.
    """

    @property
    @abstractmethod
    def model_id(self) -> str:
        """The model this judge actually invokes.

        Declared so that :func:`evaluate_judge` can check it against the
        ``JudgeConfig.model`` the policy asked for. Without it, a policy demanding one
        model would be silently judged by whatever the deployment happened to wire up
        -- the policy field would look like a control and be decoration. A verdict is
        only meaningful against the model that produced it, which is also why two
        policies naming different models refuse to merge.

        Reported by the implementation rather than configured alongside it, so the
        value comes from whatever really issues the call.
        """
        raise NotImplementedError

    @abstractmethod
    def evaluate(self, request: JudgeRequest) -> JudgeResult:
        """Evaluate a call against its purpose.

        Implementations should surface a timeout or a transport failure as a
        low-confidence result rather than a confident ``aligned=True``. The caller
        maps either to :attr:`JudgeDisposition.escalate`; a fabricated confident
        alignment is the one response that cannot be recovered from.
        """
        raise NotImplementedError


def judge_disposition(
    result: JudgeResult, config: JudgeConfig | None
) -> JudgeDisposition:
    """Decide what to do with a verdict.

    Every path that is not a confident verdict escalates, and escalation denies unless
    a handler is wired. The ordering matters: the unusable-input checks come first, so
    a malformed result cannot reach the threshold comparisons and win one.

    Three unusable-input cases, each of which would otherwise produce a confident
    allow:

    - A confidence outside ``[0, 1]``. A judge reporting ``1.5`` has malfunctioned,
      and comparing it against a threshold would grant it more authority than a
      correct answer -- ``1.5`` clears every bar.
    - Thresholds inverted, with escalation above confidence. There is no reading of
      that configuration to act on, so it escalates rather than picking whichever
      bound happens to be checked first. Merging two judge configs can produce this,
      since both thresholds take the maximum independently.
    - ``nan``, which fails every comparison and would otherwise fall through to
      whatever the last branch happened to be.
    """
    judge = config if config is not None else JudgeConfig()
    confidence_threshold = (
        DEFAULT_CONFIDENCE_THRESHOLD
        if judge.confidence_threshold is None
        else judge.confidence_threshold
    )
    escalation_threshold = (
        DEFAULT_ESCALATION_THRESHOLD
        if judge.escalation_threshold is None
        else judge.escalation_threshold
    )

    if math.isnan(result.confidence) or result.confidence < 0.0 or result.confidence > 1.0:
        return JudgeDisposition.escalate

    if escalation_threshold > confidence_threshold:
        return JudgeDisposition.escalate

    if result.confidence < escalation_threshold:
        return JudgeDisposition.escalate

    if result.confidence < confidence_threshold:
        return JudgeDisposition.escalate

    return JudgeDisposition.allow if result.aligned else JudgeDisposition.block


def _enabled_judge(
    policy: EffectivePolicy,
) -> tuple[PurposeProfile, JudgeConfig] | None:
    """The profile and judge config when this policy asks for a judge, else ``None``.

    One place that answers "is there a judge, and which one", so the callers below
    need no ``assert`` to convince themselves both objects are present. An ``assert``
    would be a branch no test can take the other side of, which is the shape of
    coverage that reads as verified and is not.

    Checked against ``is True`` rather than for truthiness, because
    ``JudgeConfig.enabled`` is ``None``-able: absent means "not configured", which is
    not the same statement as an explicit ``False`` even though both mean no judge
    today.
    """
    profile = policy.purpose_profile
    if profile is None or profile.judge is None or profile.judge.enabled is not True:
        return None
    return profile, profile.judge


def judge_enabled(policy: EffectivePolicy) -> bool:
    """Whether this policy asks for a judge at all."""
    return _enabled_judge(policy) is not None


def judge_history_window(policy: EffectivePolicy) -> int:
    """How many preceding calls the policy wants the judge to see.

    Read this to size a :class:`~tolap_core.history.ToolCallHistory`, so the window is
    the policy author's choice rather than the integrator's default. A window smaller
    than the policy asked for hides exactly the trajectory the judge was enabled to
    notice.
    """
    profile = policy.purpose_profile
    if profile is None or profile.judge is None or profile.judge.history_window is None:
        return DEFAULT_HISTORY_WINDOW
    return profile.judge.history_window


def build_judge_request(
    policy: EffectivePolicy,
    current_tool_call: str,
    history: ToolCallHistory | None = None,
) -> JudgeRequest | None:
    """Build the request for a call, from the policy's judge configuration.

    Returns ``None`` when this policy has no judge enabled -- in which case there is
    nothing to ask and the deterministic decision stands.

    The latency budget comes from the policy rather than from the caller, which is the
    whole point: ``max_latency_ms`` is a policy field and a caller passing its own
    value would make it advisory.
    """
    if current_tool_call is None:
        raise ValueError("current_tool_call must be a string, not None")

    enabled = _enabled_judge(policy)
    if enabled is None:
        return None

    profile, config = enabled
    return _request_for(profile, config, current_tool_call, history)


def _request_for(
    profile: PurposeProfile,
    config: JudgeConfig,
    current_tool_call: str,
    history: ToolCallHistory | None,
) -> JudgeRequest:
    """Assemble a request from a profile and its (present) judge configuration."""
    # Trimmed to the policy's window even when the caller's history is larger, so an
    # oversized buffer cannot quietly widen what the policy chose to send. Trimmed to
    # the MOST RECENT entries: a window of three means the last three calls.
    window = (
        DEFAULT_HISTORY_WINDOW
        if config.history_window is None
        else config.history_window
    )
    recent = history.get_recent() if history is not None else []
    if len(recent) > window:
        recent = recent[-window:]

    return JudgeRequest(
        purpose=profile,
        current_tool_call=current_tool_call,
        recent_history=recent,
        max_latency_ms=(
            DEFAULT_MAX_LATENCY_MS
            if config.max_latency_ms is None
            else config.max_latency_ms
        ),
    )


def evaluate_judge(
    policy: EffectivePolicy,
    judge: Judge,
    current_tool_call: str,
    history: ToolCallHistory | None = None,
) -> JudgeOutcome:
    """Run the judge for a call and map the verdict to a disposition and a reason.

    A policy with no judge enabled yields an allow without the judge being invoked, and
    with ``result=None`` -- so it is safe to call unconditionally. Otherwise
    ``judge.model_id`` is verified against the policy's ``model`` **before** the call:
    invoking the wrong model and then noticing would have spent the tokens and, worse,
    produced a verdict that reads as authoritative in an audit log.

    :attr:`JudgeDisposition.escalate` is **not** an allow, which is why the return value
    carries :attr:`JudgeOutcome.allowed` rather than leaving each caller to remember.
    """
    if judge is None:
        raise ValueError("evaluate_judge needs a judge; None is not a decision")
    if current_tool_call is None:
        raise ValueError("current_tool_call must be a string, not None")

    enabled = _enabled_judge(policy)
    if enabled is None:
        return JudgeOutcome(
            disposition=JudgeDisposition.allow, reason=NO_JUDGE_CONFIGURED_REASON
        )

    profile, config = enabled

    if not _model_matches(config.model, judge.model_id):
        # Named in full, so an operator reads the fault rather than inferring it. The
        # contract token stays a prefix, so `reason.startswith(JUDGE_MODEL_MISMATCH_REASON)`
        # keeps working for an integrator branching on it. `result` stays None: no model was
        # consulted, and an audit trail must be able to tell that from an uncertain verdict.
        return JudgeOutcome(
            disposition=JudgeDisposition.escalate,
            reason=(
                f"{JUDGE_MODEL_MISMATCH_REASON}: policy requires {config.model!r}, "
                f"judge is {judge.model_id!r}"
            ),
        )

    request = _request_for(profile, config, current_tool_call, history)

    try:
        result = judge.evaluate(request)
    except Exception as exc:  # noqa: BLE001 -- deliberately broad; see JUDGE_FAILED_REASON
        # `Judge` is a public ABC, so the exception taxonomy is whatever an implementation
        # happens to raise, and every one of them means the same thing here: no verdict.
        # `result` stays absent, because none was obtained.
        return JudgeOutcome(
            disposition=JudgeDisposition.escalate,
            reason=f"{JUDGE_FAILED_REASON}: {type(exc).__name__}",
        )

    disposition = judge_disposition(result, config)

    return JudgeOutcome(
        disposition=disposition,
        # The judge's own words become the reason on every disposition, so an allow records
        # why the model approved rather than leaving a reviewer to reach into `result`. An
        # audit trail reading a borderline approval wants the view, not just the verdict.
        reason=result.reasoning,
        result=result,
    )


def _model_matches(configured: str | None, actual: str) -> bool:
    """Whether the judge in hand is the one the policy asked for.

    A policy naming no model accepts any judge: the field is optional, and requiring
    it would make every judge-enabled policy fail until someone pinned a model id that
    differs per account and region.

    When a policy does name one, the comparison is case-sensitive and exact. Not a
    prefix or substring match: ``claude-sonnet`` and ``claude-sonnet-5`` are different
    models, and a prefix rule would let a deployment satisfy a policy demanding one by
    wiring the other. This mirrors the ``purpose_id`` comparison -- an identifier is
    matched, not a pattern.
    """
    return configured is None or configured == actual
